signature/cose/cose.go

```diff
@@ -15,34 +26,524 @@ func init() {
+	if err != nil {
+		return f(fmt.Sprintf("certificate-chain is invalid, %s", err))
+	}
+
+	resSignAlgo, err := getSignatureAlgorithm(certChain[0])
+	if err != nil {
+		return f(err.Error())
+	}
+	if resSignAlgo != expectedAlg {
+		return f("mismatch between signature algorithm derived from signing certificate and signing algorithm specified")
+	}
+
+	return nil
+}
+
+// getSignatureAlgorithm picks up a recommended signing algorithm for given certificate.
+func getSignatureAlgorithm(signingCert *x509.Certificate) (cose.Algorithm, error) {
+	keySpec, err := signature.ExtractKeySpec(signingCert)
+	if err != nil {
+		return 0, err
+	}
+	return getSignatureAlgorithmFromKeySpec(keySpec)
+}
+
+func getSignatureAlgorithmFromKeySpec(keySpec signature.KeySpec) (cose.Algorithm, error) {
+	switch keySpec.Type {
+	case signature.KeyTypeRSA:
+		switch keySpec.Size {
+		case 2048:
+			return signatureAlgCOSEAlgMap[signature.AlgorithmPS256], nil
+		case 3072:
+			return signatureAlgCOSEAlgMap[signature.AlgorithmPS384], nil
+		case 4096:
+			return signatureAlgCOSEAlgMap[signature.AlgorithmPS512], nil
+		default:
+			return 0, errors.New("key size not supported")
+		}
+	case signature.KeyTypeEC:
+		switch keySpec.Size {
+		case 256:
+			return signatureAlgCOSEAlgMap[signature.AlgorithmES256], nil
+		case 384:
+			return signatureAlgCOSEAlgMap[signature.AlgorithmES384], nil
+		case 512:
+			return signatureAlgCOSEAlgMap[signature.AlgorithmES512], nil
+		default:
+			return 0, errors.New("key size not supported")
+		}
+	default:
+		return 0, errors.New("key type not supported")
+	}
+}
+
+func generateCoseProtectedHeaders(req *signature.SignRequest, protected cose.ProtectedHeader) error {
+	// crit, signingScheme, expiry, signingTime, authenticSigningTime
+	var crit []interface{}
+	crit = append(crit, headerKeySigningScheme)
+	protected[headerKeySigningScheme] = string(req.SigningScheme)
+	if !req.Expiry.IsZero() {
+		crit = append(crit, headerKeyExpiry)
+		protected[headerKeyExpiry] = uint(req.Expiry.Unix())
+	}
+	switch req.SigningScheme {
+	case signature.SigningSchemeX509:
+		protected[headerKeySigningTime] = uint(req.SigningTime.Unix())
+	case signature.SigningSchemeX509SigningAuthority:
+		crit = append(crit, headerKeyAuthenticSigningTime)
+		protected[headerKeyAuthenticSigningTime] = uint(req.SigningTime.Unix())
+	default:
+		return errors.New("SigningScheme: require notary.x509 or notary.x509.signingAuthority")
+	}
+	protected[cose.HeaderLabelCritical] = crit
+
+	// content type
+	protected[cose.HeaderLabelContentType] = req.Payload.ContentType
+
+	// extended attributes
+	for _, elm := range req.ExtendedSignedAttributes {
+		if elm.Critical {
+			crit = append(crit, elm.Key)
+		}
+		protected[elm.Key] = elm.Value
+	}
+
+	return nil
+}
+
+func sign(req *signature.SignRequest, signer signature.Signer, msgToSign *cose.Sign1Message, alg cose.Algorithm) ([]byte, []*x509.Certificate, error) {
+	var coseSigner cose.Signer
+	var err error
```

Comment:
```suggestion
	var（
	    coseSigner cose.Signer
	    err error
	 ）
	    
```
I prefer defining them in a single bracket,